### DISCUSSION TITLE
Can I remove targets on iOS?

### DISCUSSION THREAD

**Discussion by dandre9** [yasirkula/UnityNativeShare#133] [Q&A] [answered]
Hi, first of all this plugin is amazing, easy to config and its working perfectly.
My only question is: is there a way to set or remove targets on iOS?

My intention is to limit share targets to only facebook, instagram and twitter

**Comment by yasirkula** [comment, marked as answer]
With a small modification, sure! https://forum.unity.com/threads/native-share-for-android-ios-open-source.519865/page-4#post-4011874

**Reply by dandre9** [reply]
Thanks for the quick response!

I dont know if I have to open another thread but one more question poped, the call NativeShare.TargetExists is not working
I only tested on Android so I dont know if iOS has the same problem
Is there an alternative to this?

**Reply by yasirkula** [reply]
But it works on Android?

**Reply by dandre9** [reply]
No, it doesnt
I found this thread:
https://forum.unity.com/threads/native-share-for-android-ios-open-source.519865/page-16#post-7632232

I think is related

**Reply by yasirkula** [reply]
That user had a different problem. TargetExists should work on Android. Can you post your code?

**Reply by dandre9** [reply]
```csharp 
IEnumerator Sharing()
{
        float widthRate = Screen.width / canvasBaseResolution.x;
        float heightRate = Screen.height / canvasBaseResolution.y;

        float imgWidth = (shareAreaRect.rect.width * widthRate);
        float imgHeight = (shareAreaRect.rect.height * heightRate);

        Vector2 temp = shareAreaRect.transform.position;
        var startX = temp.x - imgWidth / 2;
        var startY = temp.y - imgHeight / 2;

        yield return new WaitForEndOfFrame();

        Texture2D ss = new Texture2D((int)imgWidth, (int)imgHeight, TextureFormat.RGB24, false);

        Rect screenShotRect = new Rect(startX, startY, (int)imgWidth, (int)imgHeight);

        ss.ReadPixels(screenShotRect, 0, 0);
        ss.Apply();

        string filePath = Path.Combine(Application.temporaryCachePath, "radarmissionshared.png");
        File.WriteAllBytes(filePath, ss.EncodeToPNG());

        // To avoid memory leaks
        Destroy(ss);

        //Native share function
        // new NativeShare().AddFile(filePath).SetCallback((result, shareTarget) => congratsManager.onShareCallback.Invoke(result, shareTarget)).Share();

        if (NativeShare.TargetExists("com.whatsapp"))
            new NativeShare().AddFile(filePath).AddTarget("com.whatsapp").SetCallback((result, shareTarget) => congratsManager.onShareCallback.Invoke(result, shareTarget)).Share();
    }
```

I'm testing on Android 11 and the target API is set to 30
I have whatsapp installed but TargetExists is returning false
When I remove the IF statement it opens whatsapp perfectly

**Reply by yasirkula** [reply]
If you remove SetTarget, what is the value of shareTarget for WhatsApp (in the callback function)?

**Reply by dandre9** [reply]
I will check this, but just for information, I just built an APK setting the target API to 29 and now it is working

**Reply by dandre9** [reply]
I removed AddTarget, shared on Whatsapp and shareTarget returned this:

com.whatsapp/com.whatsapp.ContactPicker
MyCompany.Congrats.CongratsManager:OnShareCallback(ShareResult, String)
MyCompany.Congrats.OnShareRensponse:Invoke(ShareResult, String)
ShareResultCallback:Invoke(ShareResult, String)
NativeShareNamespace.NSCallbackHelper:Update()

**Reply by yasirkula** [reply]
Thank you for providing these details. I'll take a look at the native source code once again. BTW, did you get `unexpected element <queries> found in <manifest>` error during build before and if so, how did you resolve it?

**Reply by dandre9** [reply]
No, I didn't get this error before

Thanks for your attention!

**Reply by yasirkula** [reply]
Does AndroidManifest inside NativeShare.aar contain the `<queries>` part? Absence of it can cause this issue.

**Reply by dandre9** [reply]
Sorry for the late response, this is what is inside NativeShare.aar

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="com.yasirkula.unity">
   <application>
      <activity android:name=".NativeShareCustomShareDialogActivity" android:theme="@style/Theme.NativeShareTransparent" />
      <provider
        android:name="com.yasirkula.unity.NativeShareContentProvider"
        android:authorities="${applicationId}.NativeShareContentProvider"
        android:exported="false"
        android:grantUriPermissions="true" />
   </application>
   <uses-sdk android:targetSdkVersion="4" />
</manifest>
```

**Reply by yasirkula** [reply]
You may be using an older version of the plugin (at least your manifest is definitely old). Updating to latest version will likely resolve this issue.

**Reply by dandre9** [reply]
Yeah, that was it 
I updated and it is working fine

Sorry to worry you and thanks for all the attention

**Reply by yasirkula** [reply]
You're welcome!
